finality-aleph/src/justification.rs

```diff
@@ -245,3 +245,44 @@ where
         ))
     }
 }
+
+pub mod v1 {
+    use super::*;
+    /// Old format of justifications, needed for backwards compatibility.
+    #[derive(Clone, Encode, Decode, Debug, PartialEq)]
+    pub(crate) struct AlephJustificationV1 {
+        pub(crate) signature: SignatureSet<SignatureV1>,
+    }
+
+    impl From<AlephJustificationV1> for AlephJustification {
+        fn from(just_v1: AlephJustificationV1) -> AlephJustification {
+            let size = just_v1.signature.size();
+            let just_drop_id: SignatureSet<Signature> = just_v1
+                .signature
+                .into_iter()
+                .fold(SignatureSet::with_size(size), |sig_set, (id, sgn)| {
+                    sig_set.add_signature(&sgn.into(), id)
+                });
+            AlephJustification {
+                signature: just_drop_id,
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) enum JustificationDecoding {
+    V1(v1::AlephJustificationV1),
+    V2(AlephJustification),
+    Err,
+}
+
+pub(crate) fn backwards_compatible_decode(justification_raw: Vec<u8>) -> JustificationDecoding {
+    if let Ok(justification) = AlephJustification::decode_all(&justification_raw) {
```

Comment:
I have problems with `decode_all`
> Extension trait to Decode that ensures that the given input data is consumed completely while decoding.

If consuming means dropping the data afterwards, then I do not understand why these if-elses work. Otherwise, why standard `decode` is not sufficient?

Reply:
Normally `decode` on a bytearray `b=b1||b2` (concatenation of `b1` and `b2`) will return a success result even if only `b1` was used for the decoding (this is more complicated because of the type `decode` expects, but the idea is like that). Using `decode_all` ensures that this does not happen, this is what `consumes` means in this setting. 

I was testing this a little bit and there are cases where on some dummy data a prefix of a valid V1 encoding is a valid V2 encoding. `decode_all` prevents it -- I don't have a better solution at hand...

Reply:
thank you for the clarification!

Reply:
I mean, a hacky solution would be to have a custom encoding/decoding for `V2` = start it with a prefix like `0x13370666`: this way `V1` will never decode as `V2`, which is what we actually want :).